packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/bulk_update.ts

```diff
@@ -162,86 +206,122 @@ export const performBulkUpdate = async <T>(
-        const docFound = indexFound && isMgetDoc(actualResult) && actualResult.found;
-        if (
-          !docFound ||
+      const {
+        esRequestIndex,
+        id,
+        type,
+        version,
+        documentToSave,
+        objectNamespace,
+        migrationVersionCompatibility,
+      } = expectedBulkGetResult.value;
+
+      let namespaces: string[] | undefined;
+      const versionProperties = getExpectedVersionProperties(version);
+      const indexFound = bulkGetResponse?.statusCode !== 404;
+      const actualResult = indexFound ? bulkGetResponse?.body.docs[esRequestIndex] : undefined;
+      const docFound = indexFound && isMgetDoc(actualResult) && actualResult.found;
+      const isMultiNS = registry.isMultiNamespace(type);
+
+      if (
+        !docFound ||
+        (isMultiNS &&
           !rawDocExistsInNamespace(
             registry,
             actualResult as SavedObjectsRawDoc,
             getNamespaceId(objectNamespace)
-          )
-        ) {
-          return left({
-            id,
-            type,
-            error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
-          });
-        }
+          ))
+      ) {
+        return left({
+          id,
+          type,
+          error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
+        });
+      }
+
+      if (isMultiNS) {
         // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
         namespaces = actualResult!._source.namespaces ?? [
           // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
           SavedObjectsUtils.namespaceIdToString(actualResult!._source.namespace),
         ];
-        versionProperties = getExpectedVersionProperties(version);
-      } else {
-        if (registry.isSingleNamespace(type)) {
-          // if `objectNamespace` is undefined, fall back to `options.namespace`
-          namespaces = [getNamespaceString(objectNamespace)];
-        }
-        versionProperties = getExpectedVersionProperties(version);
+      } else if (registry.isSingleNamespace(type)) {
+        // if `objectNamespace` is undefined, fall back to `options.namespace`
+        namespaces = [getNamespaceString(objectNamespace)];
+      }
+
+      const document = getSavedObjectFromSource<T>(
+        registry,
+        type,
+        id,
+        actualResult as SavedObjectsRawDoc,
+        { migrationVersionCompatibility }
+      );
+
+      let migrated: SavedObject<T>;
+      try {
+        migrated = migrationHelper.migrateStorageDocument(document) as SavedObject<T>;
+      } catch (migrateStorageDocError) {
+        throw SavedObjectsErrorHelpers.decorateGeneralError(
+          migrateStorageDocError,
+          'Failed to migrate document to the latest version.'
+        );
       }
 
+      const typeDefinition = registry.getType(type)!;
+      const updatedAttributes = mergeForUpdate({
```

Comment:
The heart of this PR: update documents client side by merging the requested changes with the existing document to create an updated doc.